(GH-225) Add custom_insync to rsapi docs (#153)
puppetlabs/puppet-specifications: language/resource-api/README.md

      behaviour: :parameter,
      default:   false,
    },
  },
)

# lib/puppet/provider/test_custom_insync/test_custom_insync.rb
class Puppet::Provider::TestCustomInsync::TestCustomInsync < Puppet::ResourceApi::SimpleProvider
  def insync?(context, name, property_name, is_hash, should_hash)
    context.notice("Checking whether #{property_name} is out of sync")
    case property_name
    when :some_array
      if should_hash[:force]
        context.notice("Checking an order independent array")
        return is_hash[property_name].sort == should_hash[property_name].sort
      else
        context.notice("Checking a subset match array")
        found_members = (is_hash[property_name] & should_hash[property_name]).sort
        missing_members = should_hash[property_name].reject { |member| found_members.include?(member) }
        return missing_members.empty? ? true : "Adding missing members #{missing_members}"
      end
    end
end
```

When the `custom_insync` feature flag is specified, `insync?` is called once for each of the properties.
This can be used to implement custom state comparisons, when the default value-equality is not sufficient.

> The `insync?` method receives a `context` for logging, the `name` of the resource being checked, the `property_name` that is currently being inspected for insync status, the current state of the resource in `is_hash`, as well as the desired state (canonicalized) in `should_hash`.

Author: Moved to after the return value explanation.

> The `insync?` method is executed exactly once for each of the properties of the type per resource instance.

Author: Modified to read:

> The `insync?` method is executed exactly once for each of the **properties** (not parameters, read-only, or init attributes) of the type per resource instance.
> If no properties were specified in the instance of a resource, `insync?` is not called.

> The `insync?` method is executed exactly once for each of the properties of the type per resource instance.

Author: wasn't sure on the wording for read-only/init.

> The call happens during application of the catalog when puppet decides whether or not it needs to enforce state on a specific resource.

Reviewer: ```suggestion
The call happens during application of the catalog when the runtime environment decides whether or not it needs to enforce state on a specific resource.
```

The `insync?` method's return value will drive the behavior of Puppet for that property:

> - If `insync?` returns `nil`, property comparison is handed back to Puppet's built-in property insync checking.

Reviewer: ```suggestion
- If `insync?` returns `nil`, property comparison is handed back to the runtime environment's built-in property value equality checking.
```

>   - This means that _only_ properties which actually need custom insync logic need to be handled in the `insync?` method on the provider. Everything else can be ignored, which will return `nil` and use default comparison logic.

Reviewer: ```suggestion
  - This means that the provider _only_ needs to deal with properties which actually need custom insync logic. All other properties can be ignored, which will return ruby's default `nil` return value and use default comparison logic.
```

- If `insync?` returns `true`, the property is treated as insync.

> - If `insync?` returns `false`, the property is treated as out of sync and will surface the default Puppet change reporting.

Reviewer: ```suggestion
- If `insync?` returns `false`, the property is treated as out of sync and will cause the default change reporting.
```

>   - e.g. `changed <is value> to <should value>`

Reviewer: ```suggestion
  - e.g. `Notice: /Stage[main]/Main/File[/tmp/foo]/mode: mode changed <is value> to <should value>`
```

> - If `insync?` returns a string, the property is treated as out of sync and the string is surfaced in the Puppet change reporting.

Reviewer: ```suggestion
- If `insync?` returns a string, the property is treated as out of sync and the string is used for change reporting instead of the default. This can be used to provide additional information about the reason for a change.
```

>   - In the `test_custom_insync` provider example above, you would get the message `Adding missing members ['SomeValue']` if `SomeValue` was specified as a member of `some_array` but not found in the actual state returned from `get`.

Reviewer: ```suggestion
  - In the `test_custom_insync` provider example above, you would get the message `Notice: /Stage[main]/Main/Test_custom_insync[something]/some_array: Adding missing members ['SomeValue']` if `SomeValue` was specified as a member of `some_array` but not found in the actual state returned from `get`.
```

> Custom insync logic determined that this resource is out of sync

```ruby
# lib/puppet/type/test_custom_insync_hidden_property.rb
Puppet::ResourceApi.register_type(
  name: 'test_custom_insync_hidden_property',
  docs: <<-EOS,
      This type provides Puppet with the capabilities to manage ...
    EOS
  features: ['custom_insync'],
  attributes:   {
    name:        {
      type:      'String',
      desc:      'The name of the resource you want to manage.',
      behaviour: :namevar,
    },
    force: {
      type:      'Boolean',
      desc:      'If true, the resource will be treated as not being insync.',
      behaviour: :parameter,
      default:   false,
    },
  },
)

# lib/puppet/provider/test_custom_insync_hidden_property/test_custom_insync_hidden_property.rb
class Puppet::Provider::TestCustomInsyncHiddenProperty::TestCustomInsyncHiddenProperty
  def get(_context)
    [
      {
        name: 'example'
      }
    ]
  end

  def set(context, changes)
    changes.each do |_name, is_and_should|
      context.notice("Setting with #{is_and_should[:should].inspect}")
    end
  end

  def insync?(context, name, _property_name, is_hash, should_hash)

>     context.notice("Checking whether #{name} is out of sync")
> 
>     return true unless should_hash[:force]
>     context.notice("Out of sync!")

Reviewer: ```suggestion
    context.debug("Checking whether #{name} is out of sync")

    return true unless should_hash[:force]
    context.debug("Out of sync!")
```

```

### Provider feature: `custom_insync`

Allows for overriding default property `insync?` checking using custom logic in the provider.

```ruby
# lib/puppet/type/apt_key.rb
Puppet::ResourceApi.register_type(
  name: 'apt_key',
  features: [ 'custom_insync' ],
)

# lib/puppet/provider/apt_key/apt_key.rb
class Puppet::Provider::AptKey::AptKey
  def insync?(context, name, attribute_name, is_hash, should_hash)
    case attribute_name
    when :some_array_attribute

>       if should_hash[:force]

Reviewer: please do define the attributes used here in the type, so that it is clear how they are defined (property vs parameter).

Since this is also only called for properties, it's `property_name`, not `attribute_name` in the sig. Also needs to be changed in the test in https://github.com/puppetlabs/puppet-resource_api/pull/285

        context.notice("Checking an order independent array")
        # This will return true if and only if the members of both
        # arrays match entirely, regardless of order.
        return is_hash[attribute_name].sort == should_hash[attribute_name].sort
      else
        context.notice("Checking a subset match array")
        # This will return true so long as all values in the array for
        # the should_hash are also in the array of the is_hash.
        return (is_hash[attribute_name] & should_hash[attribute_name]).sort == should_hash[attribute_name].sort
      end
    end
  end
```

> Specifying the `custom_insync` feature flag ensures that all of the property attributes on the type will call the `insync?` method in the provider, if it exists, enabling custom state comparisons as complex or simple as required. The `insync?` method in the provider has access to the `context` for logging, the name of the resource being checked, the attribute that is currently being inspected for insync status, and hashes representing the current state of the resource as well as the canonicalized should state of the resource.

Reviewer: ```suggestion
When the `custom_insync` feature flag is specified, `insync?` is called once for each of the properties. This can be used to implement custom state comparisons, when the default value-equality is not sufficient. The `insync?` method receives a `context` for logging, the `name` of the resource being checked, the `property_name` that is currently being inspected for insync status, the current state of the resource in `is_hash`, as well as the desired state (canonicalized) in `should_hash`.
```

> The `insync?` method is only and always triggered for each of the property attributes of the type.

Reviewer: ```suggestion
The `insync?` method is executed exactly once for each of the properties of the type per resource instance.
The call happens during application of the catalog when puppet decides whether or not it needs to enforce state on a specific resource.
```

> If the custom `insync?` method in the provider must return a truthy or falsey result, or `nil`; if the return is `nil` the comparison defaults to the existing comparison logic for properties in Puppet. This means that unless the `insync?` method in the provider has special handling for the attribute Puppet will default to the existing state comparison behaviors. Therefore, only properties which need special handling need to be addressed in the `insync?` method in the provider.

Reviewer: ```suggestion
The `insync?` method's return value will be interpreted as truthy or falsey, following the regular ruby rules, except `nil` values; if the return is `nil` the comparison defaults to the existing comparison logic for properties in Puppet. This means that unless the `insync?` method in the provider has special handling for the attribute Puppet will default to the existing state comparison behaviors. Therefore, only properties which need special handling need to be addressed in the `insync?` method in the provider.
```

> values changed
```

### Provider feature: `custom_insync`

> Allows for overriding default property `insync?` checking using custom logic in the provider.

Reviewer 2: ```suggestion
The `custom_insyc` feature lets you for override default property `insync?` checking using custom logic in the provider.
```

      type: 'Optional[Array[String]]',
      desc: 'An array aiding array attestation',
    },
    force: {
      type:      'Boolean',
      desc:      'If true, the specified array is inclusive and must match the existing state exactly.',
      behaviour: :parameter,
      default:   false,
    },
  },
)

# lib/puppet/provider/test_custom_insync/test_custom_insync.rb
class Puppet::Provider::TestCustomInsync::TestCustomInsync < Puppet::ResourceApi::SimpleProvider
  def insync?(context, name, property_name, is_hash, should_hash)
    context.debug("Checking whether #{property_name} is out of sync")
    case property_name
    when :some_array
      if should_hash[:force]
        context.debug("Checking an order independent array")
        return is_hash[property_name].sort == should_hash[property_name].sort
      else
        context.debug("Checking a subset match array")
        found_members = (is_hash[property_name] & should_hash[property_name]).sort
        missing_members = should_hash[property_name].reject { |member| found_members.include?(member) }
        return missing_members.empty? ? true : "Adding missing members #{missing_members}"
      end
    end
end
```

> In normal operations, the runtime environment uses strict value equality to determine if a resource instance needs to be state-enforced.

Reviewer 2: Does this mean without using this do feature?

Author: It does.

Reviewer 2: Maybe say "When you don't use "custom_insync, the runtime environment etc...."?

Author: I believe I've addressed this by breaking the docs up a little to clarify:

https://github.com/puppetlabs/puppet-specifications/pull/153/files#diff-23f362101839bf8ed43d75a53178cf3cb35aa2d8169fc0163f3a7619f87ef79fR421-R424

> When the `custom_insync` feature flag is specified, `insync?` is called once for each of the properties.

Reviewer 2: ```suggestion
When you specify the `custom_insync` feature flag, `insync?` is called once for each of the properties.
```

> This can be used to implement custom state comparisons, when the default value-equality is not sufficient.

Reviewer 2: ```suggestion
You can use `custom_insync` to implement custom state comparisons when the default value-equality is not sufficient.
```

> The `insync?` method is executed exactly once for each of the **properties** (not parameters, read-only, or init attributes) of the type per resource instance.

Reviewer 2: ```suggestion
The `insync?` method is executed once for each of the **properties** (not parameters, read-only, or init attributes) of the type per resource instance.
```

If no properties were specified in the instance of a resource, `insync?` is not called.

> The call happens during application of the catalog when the runtime environment decides whether or not it needs to enforce state on a specific resource.

Reviewer 2: Is this desired state?

Author: It determines if the resource is _in_ the desired state.

Reviewer 2: Maybe rephrase it like this?

> The `insync?` method's return value will drive the behavior of Puppet for that property:

Reviewer 2: ```suggestion
The `insync?` method's return value changes the way Puppet handles the property. For example:
```

- If `insync?` returns `nil`, property comparison is handed back to the runtime environment's built-in property value equality checking.

>   - This means that the provider _only_ needs to deal with properties which actually need custom insync logic. All other properties can be ignored, which will return ruby's default `nil` return value and use default comparison logic.

Reviewer 2: Capital R for Ruby.

  - e.g. `Notice: /Stage[main]/Main/File[/tmp/foo]/mode: mode changed <is value> to <should value>`
- If `insync?` returns a string, the property is treated as out of sync and the string is used for change reporting instead of the default. This can be used to provide additional information about the reason for a change.
  - In the `test_custom_insync` provider example above, you would get the message `Notice: /Stage[main]/Main/Test_custom_insync[something]/some_array: Adding missing members ['SomeValue']` if `SomeValue` was specified as a member of `some_array` but not found in the actual state returned from `get`.
  - This enables more accurate and useful change reporting than the default is-to-should reporting surfaced by Puppet.

The `insync?(context, name, property_name, is_hash, should_hash)` method receives the following arguments:

- a `context` to access the runtime environment,
- the `name` of the resource being checked,
- the `property_name` that is currently being inspected for insync status,
- the current state of the resource in `is_hash`, and
- the desired state (canonicalized) in `should_hash`.

> For resources which specify the `custom_insync` feature flag but have _no_ defined properties the Resource API adds a spectial hidden property, `rsapi_custom_insync_trigger`, which ensures that `insync?` is still called for the resource.

Reviewer 2: ```suggestion
For resources which specify the `custom_insync` feature flag but have _no_ defined properties, the Resource API adds a spectial hidden property —`rsapi_custom_insync_trigger` —  which ensures that `insync?` is still called for the resource.
```

This hidden property is not used or seen anywhere else in the provider, only in the `insync?` method.

As with all calls to `insync?`, a custom change reporting message can be surfaced by returning a string instead of `true` or `false`.
If no custom change reporting message is returned as a string, the change reporting in Puppet (if `false` is returned) will always be:

> > Custom insync logic determined that this resource is out of sync

Reviewer 2: Is this the message the user would see?

Author: Yes.

```ruby
# lib/puppet/type/test_custom_insync_hidden_property.rb
Puppet::ResourceApi.register_type(
  name: 'test_custom_insync_hidden_property',
  docs: <<-EOS,
      This type provides Puppet with the capabilities to manage ...
    EOS
  features: ['custom_insync'],
  attributes:   {
    name:        {
      type:      'String',
      desc:      'The name of the resource you want to manage.',
      behaviour: :namevar,
    },
    force: {
      type:      'Boolean',
      desc:      'If true, the resource will be treated as not being insync.',
      behaviour: :parameter,
      default:   false,
    },
  },
)

# lib/puppet/provider/test_custom_insync_hidden_property/test_custom_insync_hidden_property.rb

> class Puppet::Provider::TestCustomInsyncHiddenProperty::TestCustomInsyncHiddenProperty

Reviewer 2: What is this example showing compared to the first one?

Author: This shows an example _without_ an insyncable property - which adds a hidden property to ensure that `insync?` is triggered.

Reviewer 2: Ah okay, maybe make that clearer?

Author: I've moved this entire section into a subheading, does that address this concern?

https://github.com/puppetlabs/puppet-specifications/pull/153/files#diff-23f362101839bf8ed43d75a53178cf3cb35aa2d8169fc0163f3a7619f87ef79fR457

Reviewer 2: Yes that is better, thanks!

      else
        context.debug("Checking a subset match array")
        found_members = (is_hash[property_name] & should_hash[property_name]).sort
        missing_members = should_hash[property_name].reject { |member| found_members.include?(member) }
        return missing_members.empty? ? true : [false, "Adding missing members #{missing_members}"]
      end
    end
end
```

When you do not specify the `custom_insync` feature flag, the runtime environment uses strict value equality to determine if a resource instance needs to be state-enforced.

When you **do** specify the `custom_insync` feature flag, `insync?` is called once for each of the properties.
You can use `custom_insync` to implement custom state comparisons when the default value-equality is not sufficient.

The `insync?` method is executed once for each of the **properties** (not parameters, read-only, or init attributes) of the type per resource instance.
If no properties were specified in the instance of a resource, `insync?` is not called.
The call happens during application of the catalog when the runtime environment decides whether or not it needs to enforce state on a specific resource.

> The `insync?` method **must** return `nil`, `true` or `false` as the result of the property comparison:

Reviewer: this is not explaining the new API requested by @joshcooper

Author: I think this is resolved by the immediately following block which explains that the message only matters if the result is actually `false`:

![image](https://user-images.githubusercontent.com/14190564/121535951-3ed94280-c9c8-11eb-9a4c-2217ea69226c.png)
